### PR TITLE
rmaps/base: add missing call to hwloc_bitmap_andnot() in bind_generic()

### PR DESCRIPTION
**Pull request by wzamazon**
In bind_generic(), there is call to hwloc_bitmap_andnot() for API < 0x20000, but the corresponding call for API >= 0x20000 is missing. This patch added it.

### CODE REVIEW

**Comment by wzamazon** [comment]
@rhc54  Can you please review?

**Comment by rhc54** [comment]
Are you sure we shouldn't actually remove the corresponding call for `hwloc < 0x20000`? Given that things seem to be working correctly with the default HWLOC (which is > 0x20000), I'm leery to make this change. Are you seeing some error that leads you to believe this addition is required?

**Comment by wzamazon** [comment]
>Given that things seem to be working correctly with the default HWLOC (which is > 0x20000),

HWLOC 2 is not working for me.

> Are you seeing some error that leads you to believe this addition is required?

Yes. On a system with HWLOC2 and 2 packages, without this change, all processes are bound to the 1st package.

**Comment by rhc54** [comment]
> HWLOC 2 is not working for me.

Sorry to hear that - it works fine for me and for the folks at AMD. We've been testing against their topologies for their web-based mapping tool.

> Yes. On a system with HWLOC2 and 2 packages, without this change, all processes are bound to the 1st package.

Well, I'll have to take your word for it. I don't offhand see why this change would be a problem, but I guess we'll find out soon enough if others complain 🤷‍♂️

**Comment by rhc54** [comment]
bot:ibm:retest

**Comment by wzamazon** [comment]
I see.

I will list the detail of the problem I encountered here, in case any one was interested.

I was using Open MPI main branch.

I tried to launch 16 processes on 2 compute instances, each 8 processes per instance.

Each of my compute instance has 2 CPUs, each CPU has 24 core.

I run the following command just to confirm the binding:
```
ompi/main/mpirun  -n 16 -N 8  --report-bindings  --hostfile /path/to/hostfile    echo "Hello, world"
```
Before the change I got:
```
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 8 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 9 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 10 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 11 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 12 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 13 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 14 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 15 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 0 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 2 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 3 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 1 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 4 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 5 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 6 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 7 bound to package[0][core:0-23]
Hello, world
```
As can be seen, all processes are bound to the 1st CPU (package).

After the change, I got:
```
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 0 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 1 bound to package[1][core:24-47]
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 2 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 3 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 5 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 4 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 6 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-1:184000] Rank 7 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 8 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 9 bound to package[1][core:24-47]
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 10 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 11 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 12 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 13 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 14 is not bound (or bound to all available processors)
[queue-p4d24xlarge-st-p4d24xlarge-2:183540] Rank 15 is not bound (or bound to all available processors)
Hello, world
....
```

**Comment by rhc54** [comment]
> Each of my compute instance has 2 CPUs, each CPU has 24 core.

You mean 2 _packages_, yes?

The pattern you show prior to the change is absolutely correct. We default to `bind-to package`, and we are binding to the entire package. Since you only specified `map-by node`, all the procs wound up on the first package - which again, is absolutely correct.

After the change, the pattern is incorrect. We bou\nd one proc to the first package, one proc to the second package (because you declare the entire first package to be consumed) and then leave all other procs unbound.

Bottom line: this change is wrong. We should remove the corresponding line from the pre-2.0 hwloc case.

**Comment by wzamazon** [comment]
>The pattern you show prior to the change is absolutely correct. We default to bind-to package, and we are binding to the entire package. Since you only specified map-by node, all the procs wound up on the first package - which again, is absolutely correct.

I see.

Now I have a different question then:

I have 8 processes on 1 node, how can I make the processes on the same node to bind-to different packages, e.g. 4 of them bind to package 0, the other 4 bind to package 1?

**Comment by rhc54** [comment]
Simplest answer is `--map-by ppr:4:package --bind-to package`. There are other ways to get there as well, but this will do what you want.

**Comment by wzamazon** [comment]
Thank you!

Closing this one.
